// @tanstack/vue-form/dist/esm/useField.js
import { FieldApi } from "@tanstack/form-core";
import { useStore } from "@tanstack/vue-store";
import { defineComponent, computed, onMounted, onUnmounted, watch } from "vue";
function useField(opts) {
  const fieldApi = (() => {
    const api = new FieldApi({
      ...opts,
      form: opts.form,
      name: opts.name
    });
    return api;
  })();
  const reactiveStateValue = useStore(
    fieldApi.store,
    opts.mode === "array" ? (state) => state.meta._arrayVersion || 0 : (state) => state.value
  );
  const reactiveMetaIsTouched = useStore(
    fieldApi.store,
    (state) => state.meta.isTouched
  );
  const reactiveMetaIsBlurred = useStore(
    fieldApi.store,
    (state) => state.meta.isBlurred
  );
  const reactiveMetaIsDirty = useStore(
    fieldApi.store,
    (state) => state.meta.isDirty
  );
  const reactiveMetaErrorMap = useStore(
    fieldApi.store,
    (state) => state.meta.errorMap
  );
  const reactiveMetaErrorSourceMap = useStore(
    fieldApi.store,
    (state) => state.meta.errorSourceMap
  );
  const reactiveMetaIsValidating = useStore(
    fieldApi.store,
    (state) => state.meta.isValidating
  );
  const fieldState = computed(() => {
    const trackedValue = reactiveStateValue.value;
    const isTouched = reactiveMetaIsTouched.value;
    const isBlurred = reactiveMetaIsBlurred.value;
    const isDirty = reactiveMetaIsDirty.value;
    const errorMap = reactiveMetaErrorMap.value;
    const errorSourceMap = reactiveMetaErrorSourceMap.value;
    const isValidating = reactiveMetaIsValidating.value;
    return {
      value: opts.mode === "array" ? fieldApi.state.value : trackedValue,
      meta: {
        ...fieldApi.state.meta,
        isTouched,
        isBlurred,
        isDirty,
        errorMap,
        errorSourceMap,
        isValidating
      }
    };
  });
  const extendedFieldApi = computed(() => {
    const reactiveFieldApi = {
      ...fieldApi,
      get state() {
        return fieldState.value;
      }
    };
    const extendedApi = reactiveFieldApi;
    return extendedApi;
  });
  let cleanup;
  onMounted(() => {
    cleanup = fieldApi.mount();
  });
  onUnmounted(() => {
    cleanup();
  });
  watch(
    () => opts,
    () => {
      fieldApi.update({ ...opts, form: opts.form });
    }
  );
  return { api: extendedFieldApi.value, state: fieldState.value };
}
const Field = defineComponent(
  (fieldOptions, context) => {
    const fieldApi = useField({ ...fieldOptions, ...context.attrs });
    return () => context.slots.default({
      field: fieldApi.api,
      state: fieldApi.state
    });
  },
  { name: "Field", inheritAttrs: false }
);
export {
  Field,
  useField
};
