// @tanstack/vue-form/dist/cjs/useField.cjs
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const formCore = require("@tanstack/form-core");
const vueStore = require("@tanstack/vue-store");
const vue = require("vue");
function useField(opts) {
  const fieldApi = (() => {
    const api = new formCore.FieldApi({
      ...opts,
      form: opts.form,
      name: opts.name
    });
    return api;
  })();
  const reactiveStateValue = vueStore.useStore(
    fieldApi.store,
    opts.mode === "array" ? (state) => state.meta._arrayVersion || 0 : (state) => state.value
  );
  const reactiveMetaIsTouched = vueStore.useStore(
    fieldApi.store,
    (state) => state.meta.isTouched
  );
  const reactiveMetaIsBlurred = vueStore.useStore(
    fieldApi.store,
    (state) => state.meta.isBlurred
  );
  const reactiveMetaIsDirty = vueStore.useStore(
    fieldApi.store,
    (state) => state.meta.isDirty
  );
  const reactiveMetaErrorMap = vueStore.useStore(
    fieldApi.store,
    (state) => state.meta.errorMap
  );
  const reactiveMetaErrorSourceMap = vueStore.useStore(
    fieldApi.store,
    (state) => state.meta.errorSourceMap
  );
  const reactiveMetaIsValidating = vueStore.useStore(
    fieldApi.store,
    (state) => state.meta.isValidating
  );
  const fieldState = vue.computed(() => {
    const trackedValue = reactiveStateValue.value;
    const isTouched = reactiveMetaIsTouched.value;
    const isBlurred = reactiveMetaIsBlurred.value;
    const isDirty = reactiveMetaIsDirty.value;
    const errorMap = reactiveMetaErrorMap.value;
    const errorSourceMap = reactiveMetaErrorSourceMap.value;
    const isValidating = reactiveMetaIsValidating.value;
    return {
      value: opts.mode === "array" ? fieldApi.state.value : trackedValue,
      meta: {
        ...fieldApi.state.meta,
        isTouched,
        isBlurred,
        isDirty,
        errorMap,
        errorSourceMap,
        isValidating
      }
    };
  });
  const extendedFieldApi = vue.computed(() => {
    const reactiveFieldApi = {
      ...fieldApi,
      get state() {
        return fieldState.value;
      }
    };
    const extendedApi = reactiveFieldApi;
    return extendedApi;
  });
  let cleanup;
  vue.onMounted(() => {
    cleanup = fieldApi.mount();
  });
  vue.onUnmounted(() => {
    cleanup();
  });
  vue.watch(
    () => opts,
    () => {
      fieldApi.update({ ...opts, form: opts.form });
    }
  );
  return { api: extendedFieldApi.value, state: fieldState.value };
}
const Field = vue.defineComponent(
  (fieldOptions, context) => {
    const fieldApi = useField({ ...fieldOptions, ...context.attrs });
    return () => context.slots.default({
      field: fieldApi.api,
      state: fieldApi.state
    });
  },
  { name: "Field", inheritAttrs: false }
);
exports.Field = Field;
exports.useField = useField;
