// @liuli-util/fs-extra/dist/index.js
// src/index.ts
import fsExtra from "fs-extra";
var appendFile = fsExtra.appendFile;
var appendFileSync = fsExtra.appendFileSync;
var access = fsExtra.access;
var accessSync = fsExtra.accessSync;
var chown = fsExtra.chown;
var chownSync = fsExtra.chownSync;
var chmod = fsExtra.chmod;
var chmodSync = fsExtra.chmodSync;
var close = fsExtra.close;
var closeSync = fsExtra.closeSync;
var copyFile = fsExtra.copyFile;
var copyFileSync = fsExtra.copyFileSync;
var cp = fsExtra.cp;
var cpSync = fsExtra.cpSync;
var createReadStream = fsExtra.createReadStream;
var createWriteStream = fsExtra.createWriteStream;
var exists = fsExtra.exists;
var existsSync = fsExtra.existsSync;
var fchown = fsExtra.fchown;
var fchownSync = fsExtra.fchownSync;
var fchmod = fsExtra.fchmod;
var fchmodSync = fsExtra.fchmodSync;
var fdatasync = fsExtra.fdatasync;
var fdatasyncSync = fsExtra.fdatasyncSync;
var fstat = fsExtra.fstat;
var fstatSync = fsExtra.fstatSync;
var fsync = fsExtra.fsync;
var fsyncSync = fsExtra.fsyncSync;
var ftruncate = fsExtra.ftruncate;
var ftruncateSync = fsExtra.ftruncateSync;
var futimes = fsExtra.futimes;
var futimesSync = fsExtra.futimesSync;
var lchown = fsExtra.lchown;
var lchownSync = fsExtra.lchownSync;
var lchmod = fsExtra.lchmod;
var lchmodSync = fsExtra.lchmodSync;
var link = fsExtra.link;
var linkSync = fsExtra.linkSync;
var lstat = fsExtra.lstat;
var lstatSync = fsExtra.lstatSync;
var lutimes = fsExtra.lutimes;
var lutimesSync = fsExtra.lutimesSync;
var mkdir = fsExtra.mkdir;
var mkdirSync = fsExtra.mkdirSync;
var mkdtemp = fsExtra.mkdtemp;
var mkdtempSync = fsExtra.mkdtempSync;
var open = fsExtra.open;
var openSync = fsExtra.openSync;
var opendir = fsExtra.opendir;
var opendirSync = fsExtra.opendirSync;
var readdir = fsExtra.readdir;
var readdirSync = fsExtra.readdirSync;
var read = fsExtra.read;
var readSync = fsExtra.readSync;
var readv = fsExtra.readv;
var readvSync = fsExtra.readvSync;
var readFile = fsExtra.readFile;
var readFileSync = fsExtra.readFileSync;
var readlink = fsExtra.readlink;
var readlinkSync = fsExtra.readlinkSync;
var realpath = fsExtra.realpath;
var realpathSync = fsExtra.realpathSync;
var rename = fsExtra.rename;
var renameSync = fsExtra.renameSync;
var rm = fsExtra.rm;
var rmSync = fsExtra.rmSync;
var rmdir = fsExtra.rmdir;
var rmdirSync = fsExtra.rmdirSync;
var stat = fsExtra.stat;
var statSync = fsExtra.statSync;
var symlink = fsExtra.symlink;
var symlinkSync = fsExtra.symlinkSync;
var truncate = fsExtra.truncate;
var truncateSync = fsExtra.truncateSync;
var unwatchFile = fsExtra.unwatchFile;
var unlink = fsExtra.unlink;
var unlinkSync = fsExtra.unlinkSync;
var utimes = fsExtra.utimes;
var utimesSync = fsExtra.utimesSync;
var watch = fsExtra.watch;
var watchFile = fsExtra.watchFile;
var writeFile = fsExtra.writeFile;
var writeFileSync = fsExtra.writeFileSync;
var write = fsExtra.write;
var writeSync = fsExtra.writeSync;
var writev = fsExtra.writev;
var writevSync = fsExtra.writevSync;
var Dir = fsExtra.Dir;
var Dirent = fsExtra.Dirent;
var Stats = fsExtra.Stats;
var ReadStream = fsExtra.ReadStream;
var WriteStream = fsExtra.WriteStream;
var constants = fsExtra.constants;
var promises = fsExtra.promises;
var copy = fsExtra.copy;
var copySync = fsExtra.copySync;
var emptyDirSync = fsExtra.emptyDirSync;
var emptydirSync = fsExtra.emptydirSync;
var emptyDir = fsExtra.emptyDir;
var emptydir = fsExtra.emptydir;
var createFile = fsExtra.createFile;
var createFileSync = fsExtra.createFileSync;
var ensureFile = fsExtra.ensureFile;
var ensureFileSync = fsExtra.ensureFileSync;
var createLink = fsExtra.createLink;
var createLinkSync = fsExtra.createLinkSync;
var ensureLink = fsExtra.ensureLink;
var ensureLinkSync = fsExtra.ensureLinkSync;
var createSymlink = fsExtra.createSymlink;
var createSymlinkSync = fsExtra.createSymlinkSync;
var ensureSymlink = fsExtra.ensureSymlink;
var ensureSymlinkSync = fsExtra.ensureSymlinkSync;
var readJson = fsExtra.readJson;
var readJsonSync = fsExtra.readJsonSync;
var writeJson = fsExtra.writeJson;
var writeJsonSync = fsExtra.writeJsonSync;
var outputJson = fsExtra.outputJson;
var outputJsonSync = fsExtra.outputJsonSync;
var outputJSON = fsExtra.outputJSON;
var outputJSONSync = fsExtra.outputJSONSync;
var writeJSON = fsExtra.writeJSON;
var writeJSONSync = fsExtra.writeJSONSync;
var readJSON = fsExtra.readJSON;
var readJSONSync = fsExtra.readJSONSync;
var mkdirs = fsExtra.mkdirs;
var mkdirsSync = fsExtra.mkdirsSync;
var mkdirp = fsExtra.mkdirp;
var mkdirpSync = fsExtra.mkdirpSync;
var ensureDir = fsExtra.ensureDir;
var ensureDirSync = fsExtra.ensureDirSync;
var move = fsExtra.move;
var moveSync = fsExtra.moveSync;
var outputFile = fsExtra.outputFile;
var outputFileSync = fsExtra.outputFileSync;
var pathExists = fsExtra.pathExists;
var pathExistsSync = fsExtra.pathExistsSync;
var remove = fsExtra.remove;
var removeSync = fsExtra.removeSync;
export {
  Dir,
  Dirent,
  ReadStream,
  Stats,
  WriteStream,
  access,
  accessSync,
  appendFile,
  appendFileSync,
  chmod,
  chmodSync,
  chown,
  chownSync,
  close,
  closeSync,
  constants,
  copy,
  copyFile,
  copyFileSync,
  copySync,
  cp,
  cpSync,
  createFile,
  createFileSync,
  createLink,
  createLinkSync,
  createReadStream,
  createSymlink,
  createSymlinkSync,
  createWriteStream,
  emptyDir,
  emptyDirSync,
  emptydir,
  emptydirSync,
  ensureDir,
  ensureDirSync,
  ensureFile,
  ensureFileSync,
  ensureLink,
  ensureLinkSync,
  ensureSymlink,
  ensureSymlinkSync,
  exists,
  existsSync,
  fchmod,
  fchmodSync,
  fchown,
  fchownSync,
  fdatasync,
  fdatasyncSync,
  fstat,
  fstatSync,
  fsync,
  fsyncSync,
  ftruncate,
  ftruncateSync,
  futimes,
  futimesSync,
  lchmod,
  lchmodSync,
  lchown,
  lchownSync,
  link,
  linkSync,
  lstat,
  lstatSync,
  lutimes,
  lutimesSync,
  mkdir,
  mkdirSync,
  mkdirp,
  mkdirpSync,
  mkdirs,
  mkdirsSync,
  mkdtemp,
  mkdtempSync,
  move,
  moveSync,
  open,
  openSync,
  opendir,
  opendirSync,
  outputFile,
  outputFileSync,
  outputJSON,
  outputJSONSync,
  outputJson,
  outputJsonSync,
  pathExists,
  pathExistsSync,
  promises,
  read,
  readFile,
  readFileSync,
  readJSON,
  readJSONSync,
  readJson,
  readJsonSync,
  readSync,
  readdir,
  readdirSync,
  readlink,
  readlinkSync,
  readv,
  readvSync,
  realpath,
  realpathSync,
  remove,
  removeSync,
  rename,
  renameSync,
  rm,
  rmSync,
  rmdir,
  rmdirSync,
  stat,
  statSync,
  symlink,
  symlinkSync,
  truncate,
  truncateSync,
  unlink,
  unlinkSync,
  unwatchFile,
  utimes,
  utimesSync,
  watch,
  watchFile,
  write,
  writeFile,
  writeFileSync,
  writeJSON,
  writeJSONSync,
  writeJson,
  writeJsonSync,
  writeSync,
  writev,
  writevSync
};
