// @liuli-util/fs-extra/dist/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  Dir: () => Dir,
  Dirent: () => Dirent,
  ReadStream: () => ReadStream,
  Stats: () => Stats,
  WriteStream: () => WriteStream,
  access: () => access,
  accessSync: () => accessSync,
  appendFile: () => appendFile,
  appendFileSync: () => appendFileSync,
  chmod: () => chmod,
  chmodSync: () => chmodSync,
  chown: () => chown,
  chownSync: () => chownSync,
  close: () => close,
  closeSync: () => closeSync,
  constants: () => constants,
  copy: () => copy,
  copyFile: () => copyFile,
  copyFileSync: () => copyFileSync,
  copySync: () => copySync,
  cp: () => cp,
  cpSync: () => cpSync,
  createFile: () => createFile,
  createFileSync: () => createFileSync,
  createLink: () => createLink,
  createLinkSync: () => createLinkSync,
  createReadStream: () => createReadStream,
  createSymlink: () => createSymlink,
  createSymlinkSync: () => createSymlinkSync,
  createWriteStream: () => createWriteStream,
  emptyDir: () => emptyDir,
  emptyDirSync: () => emptyDirSync,
  emptydir: () => emptydir,
  emptydirSync: () => emptydirSync,
  ensureDir: () => ensureDir,
  ensureDirSync: () => ensureDirSync,
  ensureFile: () => ensureFile,
  ensureFileSync: () => ensureFileSync,
  ensureLink: () => ensureLink,
  ensureLinkSync: () => ensureLinkSync,
  ensureSymlink: () => ensureSymlink,
  ensureSymlinkSync: () => ensureSymlinkSync,
  exists: () => exists,
  existsSync: () => existsSync,
  fchmod: () => fchmod,
  fchmodSync: () => fchmodSync,
  fchown: () => fchown,
  fchownSync: () => fchownSync,
  fdatasync: () => fdatasync,
  fdatasyncSync: () => fdatasyncSync,
  fstat: () => fstat,
  fstatSync: () => fstatSync,
  fsync: () => fsync,
  fsyncSync: () => fsyncSync,
  ftruncate: () => ftruncate,
  ftruncateSync: () => ftruncateSync,
  futimes: () => futimes,
  futimesSync: () => futimesSync,
  lchmod: () => lchmod,
  lchmodSync: () => lchmodSync,
  lchown: () => lchown,
  lchownSync: () => lchownSync,
  link: () => link,
  linkSync: () => linkSync,
  lstat: () => lstat,
  lstatSync: () => lstatSync,
  lutimes: () => lutimes,
  lutimesSync: () => lutimesSync,
  mkdir: () => mkdir,
  mkdirSync: () => mkdirSync,
  mkdirp: () => mkdirp,
  mkdirpSync: () => mkdirpSync,
  mkdirs: () => mkdirs,
  mkdirsSync: () => mkdirsSync,
  mkdtemp: () => mkdtemp,
  mkdtempSync: () => mkdtempSync,
  move: () => move,
  moveSync: () => moveSync,
  open: () => open,
  openSync: () => openSync,
  opendir: () => opendir,
  opendirSync: () => opendirSync,
  outputFile: () => outputFile,
  outputFileSync: () => outputFileSync,
  outputJSON: () => outputJSON,
  outputJSONSync: () => outputJSONSync,
  outputJson: () => outputJson,
  outputJsonSync: () => outputJsonSync,
  pathExists: () => pathExists,
  pathExistsSync: () => pathExistsSync,
  promises: () => promises,
  read: () => read,
  readFile: () => readFile,
  readFileSync: () => readFileSync,
  readJSON: () => readJSON,
  readJSONSync: () => readJSONSync,
  readJson: () => readJson,
  readJsonSync: () => readJsonSync,
  readSync: () => readSync,
  readdir: () => readdir,
  readdirSync: () => readdirSync,
  readlink: () => readlink,
  readlinkSync: () => readlinkSync,
  readv: () => readv,
  readvSync: () => readvSync,
  realpath: () => realpath,
  realpathSync: () => realpathSync,
  remove: () => remove,
  removeSync: () => removeSync,
  rename: () => rename,
  renameSync: () => renameSync,
  rm: () => rm,
  rmSync: () => rmSync,
  rmdir: () => rmdir,
  rmdirSync: () => rmdirSync,
  stat: () => stat,
  statSync: () => statSync,
  symlink: () => symlink,
  symlinkSync: () => symlinkSync,
  truncate: () => truncate,
  truncateSync: () => truncateSync,
  unlink: () => unlink,
  unlinkSync: () => unlinkSync,
  unwatchFile: () => unwatchFile,
  utimes: () => utimes,
  utimesSync: () => utimesSync,
  watch: () => watch,
  watchFile: () => watchFile,
  write: () => write,
  writeFile: () => writeFile,
  writeFileSync: () => writeFileSync,
  writeJSON: () => writeJSON,
  writeJSONSync: () => writeJSONSync,
  writeJson: () => writeJson,
  writeJsonSync: () => writeJsonSync,
  writeSync: () => writeSync,
  writev: () => writev,
  writevSync: () => writevSync
});
module.exports = __toCommonJS(src_exports);
var import_fs_extra = __toESM(require("fs-extra"), 1);
var appendFile = import_fs_extra.default.appendFile;
var appendFileSync = import_fs_extra.default.appendFileSync;
var access = import_fs_extra.default.access;
var accessSync = import_fs_extra.default.accessSync;
var chown = import_fs_extra.default.chown;
var chownSync = import_fs_extra.default.chownSync;
var chmod = import_fs_extra.default.chmod;
var chmodSync = import_fs_extra.default.chmodSync;
var close = import_fs_extra.default.close;
var closeSync = import_fs_extra.default.closeSync;
var copyFile = import_fs_extra.default.copyFile;
var copyFileSync = import_fs_extra.default.copyFileSync;
var cp = import_fs_extra.default.cp;
var cpSync = import_fs_extra.default.cpSync;
var createReadStream = import_fs_extra.default.createReadStream;
var createWriteStream = import_fs_extra.default.createWriteStream;
var exists = import_fs_extra.default.exists;
var existsSync = import_fs_extra.default.existsSync;
var fchown = import_fs_extra.default.fchown;
var fchownSync = import_fs_extra.default.fchownSync;
var fchmod = import_fs_extra.default.fchmod;
var fchmodSync = import_fs_extra.default.fchmodSync;
var fdatasync = import_fs_extra.default.fdatasync;
var fdatasyncSync = import_fs_extra.default.fdatasyncSync;
var fstat = import_fs_extra.default.fstat;
var fstatSync = import_fs_extra.default.fstatSync;
var fsync = import_fs_extra.default.fsync;
var fsyncSync = import_fs_extra.default.fsyncSync;
var ftruncate = import_fs_extra.default.ftruncate;
var ftruncateSync = import_fs_extra.default.ftruncateSync;
var futimes = import_fs_extra.default.futimes;
var futimesSync = import_fs_extra.default.futimesSync;
var lchown = import_fs_extra.default.lchown;
var lchownSync = import_fs_extra.default.lchownSync;
var lchmod = import_fs_extra.default.lchmod;
var lchmodSync = import_fs_extra.default.lchmodSync;
var link = import_fs_extra.default.link;
var linkSync = import_fs_extra.default.linkSync;
var lstat = import_fs_extra.default.lstat;
var lstatSync = import_fs_extra.default.lstatSync;
var lutimes = import_fs_extra.default.lutimes;
var lutimesSync = import_fs_extra.default.lutimesSync;
var mkdir = import_fs_extra.default.mkdir;
var mkdirSync = import_fs_extra.default.mkdirSync;
var mkdtemp = import_fs_extra.default.mkdtemp;
var mkdtempSync = import_fs_extra.default.mkdtempSync;
var open = import_fs_extra.default.open;
var openSync = import_fs_extra.default.openSync;
var opendir = import_fs_extra.default.opendir;
var opendirSync = import_fs_extra.default.opendirSync;
var readdir = import_fs_extra.default.readdir;
var readdirSync = import_fs_extra.default.readdirSync;
var read = import_fs_extra.default.read;
var readSync = import_fs_extra.default.readSync;
var readv = import_fs_extra.default.readv;
var readvSync = import_fs_extra.default.readvSync;
var readFile = import_fs_extra.default.readFile;
var readFileSync = import_fs_extra.default.readFileSync;
var readlink = import_fs_extra.default.readlink;
var readlinkSync = import_fs_extra.default.readlinkSync;
var realpath = import_fs_extra.default.realpath;
var realpathSync = import_fs_extra.default.realpathSync;
var rename = import_fs_extra.default.rename;
var renameSync = import_fs_extra.default.renameSync;
var rm = import_fs_extra.default.rm;
var rmSync = import_fs_extra.default.rmSync;
var rmdir = import_fs_extra.default.rmdir;
var rmdirSync = import_fs_extra.default.rmdirSync;
var stat = import_fs_extra.default.stat;
var statSync = import_fs_extra.default.statSync;
var symlink = import_fs_extra.default.symlink;
var symlinkSync = import_fs_extra.default.symlinkSync;
var truncate = import_fs_extra.default.truncate;
var truncateSync = import_fs_extra.default.truncateSync;
var unwatchFile = import_fs_extra.default.unwatchFile;
var unlink = import_fs_extra.default.unlink;
var unlinkSync = import_fs_extra.default.unlinkSync;
var utimes = import_fs_extra.default.utimes;
var utimesSync = import_fs_extra.default.utimesSync;
var watch = import_fs_extra.default.watch;
var watchFile = import_fs_extra.default.watchFile;
var writeFile = import_fs_extra.default.writeFile;
var writeFileSync = import_fs_extra.default.writeFileSync;
var write = import_fs_extra.default.write;
var writeSync = import_fs_extra.default.writeSync;
var writev = import_fs_extra.default.writev;
var writevSync = import_fs_extra.default.writevSync;
var Dir = import_fs_extra.default.Dir;
var Dirent = import_fs_extra.default.Dirent;
var Stats = import_fs_extra.default.Stats;
var ReadStream = import_fs_extra.default.ReadStream;
var WriteStream = import_fs_extra.default.WriteStream;
var constants = import_fs_extra.default.constants;
var promises = import_fs_extra.default.promises;
var copy = import_fs_extra.default.copy;
var copySync = import_fs_extra.default.copySync;
var emptyDirSync = import_fs_extra.default.emptyDirSync;
var emptydirSync = import_fs_extra.default.emptydirSync;
var emptyDir = import_fs_extra.default.emptyDir;
var emptydir = import_fs_extra.default.emptydir;
var createFile = import_fs_extra.default.createFile;
var createFileSync = import_fs_extra.default.createFileSync;
var ensureFile = import_fs_extra.default.ensureFile;
var ensureFileSync = import_fs_extra.default.ensureFileSync;
var createLink = import_fs_extra.default.createLink;
var createLinkSync = import_fs_extra.default.createLinkSync;
var ensureLink = import_fs_extra.default.ensureLink;
var ensureLinkSync = import_fs_extra.default.ensureLinkSync;
var createSymlink = import_fs_extra.default.createSymlink;
var createSymlinkSync = import_fs_extra.default.createSymlinkSync;
var ensureSymlink = import_fs_extra.default.ensureSymlink;
var ensureSymlinkSync = import_fs_extra.default.ensureSymlinkSync;
var readJson = import_fs_extra.default.readJson;
var readJsonSync = import_fs_extra.default.readJsonSync;
var writeJson = import_fs_extra.default.writeJson;
var writeJsonSync = import_fs_extra.default.writeJsonSync;
var outputJson = import_fs_extra.default.outputJson;
var outputJsonSync = import_fs_extra.default.outputJsonSync;
var outputJSON = import_fs_extra.default.outputJSON;
var outputJSONSync = import_fs_extra.default.outputJSONSync;
var writeJSON = import_fs_extra.default.writeJSON;
var writeJSONSync = import_fs_extra.default.writeJSONSync;
var readJSON = import_fs_extra.default.readJSON;
var readJSONSync = import_fs_extra.default.readJSONSync;
var mkdirs = import_fs_extra.default.mkdirs;
var mkdirsSync = import_fs_extra.default.mkdirsSync;
var mkdirp = import_fs_extra.default.mkdirp;
var mkdirpSync = import_fs_extra.default.mkdirpSync;
var ensureDir = import_fs_extra.default.ensureDir;
var ensureDirSync = import_fs_extra.default.ensureDirSync;
var move = import_fs_extra.default.move;
var moveSync = import_fs_extra.default.moveSync;
var outputFile = import_fs_extra.default.outputFile;
var outputFileSync = import_fs_extra.default.outputFileSync;
var pathExists = import_fs_extra.default.pathExists;
var pathExistsSync = import_fs_extra.default.pathExistsSync;
var remove = import_fs_extra.default.remove;
var removeSync = import_fs_extra.default.removeSync;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Dir,
  Dirent,
  ReadStream,
  Stats,
  WriteStream,
  access,
  accessSync,
  appendFile,
  appendFileSync,
  chmod,
  chmodSync,
  chown,
  chownSync,
  close,
  closeSync,
  constants,
  copy,
  copyFile,
  copyFileSync,
  copySync,
  cp,
  cpSync,
  createFile,
  createFileSync,
  createLink,
  createLinkSync,
  createReadStream,
  createSymlink,
  createSymlinkSync,
  createWriteStream,
  emptyDir,
  emptyDirSync,
  emptydir,
  emptydirSync,
  ensureDir,
  ensureDirSync,
  ensureFile,
  ensureFileSync,
  ensureLink,
  ensureLinkSync,
  ensureSymlink,
  ensureSymlinkSync,
  exists,
  existsSync,
  fchmod,
  fchmodSync,
  fchown,
  fchownSync,
  fdatasync,
  fdatasyncSync,
  fstat,
  fstatSync,
  fsync,
  fsyncSync,
  ftruncate,
  ftruncateSync,
  futimes,
  futimesSync,
  lchmod,
  lchmodSync,
  lchown,
  lchownSync,
  link,
  linkSync,
  lstat,
  lstatSync,
  lutimes,
  lutimesSync,
  mkdir,
  mkdirSync,
  mkdirp,
  mkdirpSync,
  mkdirs,
  mkdirsSync,
  mkdtemp,
  mkdtempSync,
  move,
  moveSync,
  open,
  openSync,
  opendir,
  opendirSync,
  outputFile,
  outputFileSync,
  outputJSON,
  outputJSONSync,
  outputJson,
  outputJsonSync,
  pathExists,
  pathExistsSync,
  promises,
  read,
  readFile,
  readFileSync,
  readJSON,
  readJSONSync,
  readJson,
  readJsonSync,
  readSync,
  readdir,
  readdirSync,
  readlink,
  readlinkSync,
  readv,
  readvSync,
  realpath,
  realpathSync,
  remove,
  removeSync,
  rename,
  renameSync,
  rm,
  rmSync,
  rmdir,
  rmdirSync,
  stat,
  statSync,
  symlink,
  symlinkSync,
  truncate,
  truncateSync,
  unlink,
  unlinkSync,
  unwatchFile,
  utimes,
  utimesSync,
  watch,
  watchFile,
  write,
  writeFile,
  writeFileSync,
  writeJSON,
  writeJSONSync,
  writeJson,
  writeJsonSync,
  writeSync,
  writev,
  writevSync
});
